Proposal: Rationalizing Sequence end-operation names

brentdax opened the pull request:
This proposal suggests renaming or redesigning many APIs which operate on the beginning or end of a Sequence or Collection, with the goal of producing coherently-named method families. In summary:

* `c.drop…` becomes `c.removing…`
* `c.…First/Last(n)` become `c.…Prefix/Suffix(n)`
* `c.starts(with: prefix…)` becomes `c.hasPrefix(prefix…)`
* `c.index(of: elem/where: isCond)` becomes `c.firstIndex(of: elem/where: isCond)`
* `c.prefix(upTo: i)`, `c.prefix(through: i)`, and `c.suffix(from: i)` become `c[..<i]`, `c[...i]`, and `c[i..<]`

To support that last change, this proposal includes new `IncompleteRange` and `IncompleteClosedRange` types and operators to construct them; it is otherwise not additive.

(@dabrahams suggested reducing the number of subscript overloads using a `RangeExpression` protocol, but when writing a prototype I realized this would require generic subscripts. Next year!)

dabrahams commented:
Hi Brent,

Thanks for putting this together.  I'm sorry I wasn't able to respond
earlier to this work; we've been swamped!

on Tue Jul 19 2016, Brent Royal-Gordon <notifications-AT-github.com> wrote:

> This proposal suggests renaming or redesigning many APIs which operate
> on the beginning or end of a Sequence or Collection, with the goal of
> producing coherently-named method families. In summary:
>
> * `c.drop…` becomes `c.removing…`
> * `c.…First/Last(n)` become `c.…Prefix/Suffix(n)`
> * `c.starts(with: prefix…)` becomes `c.hasPrefix(prefix…)`
> * `c.index(of: elem/where: isCond)` becomes `c.firstIndex(of: elem/where: isCond)`
> * `c.prefix(upTo: i)`, `c.prefix(through: i)`, and `c.suffix(from: i)` become `c[..<i]`, `c[...i]`, and `c[i..<]`
>
> To support that last change, this proposal includes new `IncompleteRange` and `IncompleteClosedRange` types and operators to construct them; it is otherwise not additive.
>
> (@dabrahams suggested reducing the number of subscript overloads using
> a `RangeExpression` protocol, but when writing a prototype I realized
> this would require generic subscripts. 

It wouldn't, actually.  You'd just need to write out all the subscript
overloads, as today.  Other APIs could take advantage of it.  The main
reason I want this protocol is that I expect to see more models of
RangeExpression, such as e.g. `c[$+256..<]` (drop the first 256
elements).

> Next year!)  You can view, comment on, or merge this pull request
> online at:
>
>   https://github.com/apple/swift-evolution/pull/447
>
> -- Commit Summary --
>
>   * Proposal: Rationalizing Sequence end-operation names
>
> -- File Changes --
>
>     A proposals/NNNN-sequence-end-ops.md (766)
>
> -- Patch Links --
>
> https://github.com/apple/swift-evolution/pull/447.patch
> https://github.com/apple/swift-evolution/pull/447.diff
>
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/apple/swift-evolution/pull/447

-- 
Dave

brentdax commented:
Okay, I can try to work that into the prototype and proposal tonight. (Question: Should `Range` conform to `RangeExpression`? The rule favoring non-generics ought to prefer the `Range` overload, so it wouldn't actually be ambiguous or recursive.)

Is there anything else that should be done to this proposal?

brentdax commented:
I notice that `AnyRangeExpression<Bound>` could be a trivial closure-based type erasing wrapper. Would it be worth including?

lattner commented:
@brentdax Please send me a direct email when this is ready to merge, thanks.

brentdax commented:
Sorry to do tech support on a Github issue, but I'm stuck on a compiler error. This code:

```swift
public protocol RangeExpression {
    associatedtype Bound : Comparable
    func relative<C: Indexable>(to collection: C) -> Range<Bound> where C.Index == Bound
}
extension CountableRange: RangeExpression {
    public func relative<C: Indexable>(to collection: C) -> Range<Bound> where C.Index == Bound {
        return Range(self)
    }
}
```

Gives me: `<unknown>:0: error: same-type constraint type 'τ_0_0' does not conform to required protocol 'Comparable'`. Despite the less-than-helpful symbol, I suspect this is trying to say that, in the implementation of `relative(to:)`, the compiler thinks `C.Index` isn't constrained to be `Comparable`. Except...well, it is. Even `IndexableBase.Index` is constrained to `Comparable`.

Is this a bug? Any idea how I can work around it?

dabrahams commented:
on Wed Jul 20 2016, Brent Royal-Gordon <notifications-AT-github.com> wrote:

> I notice that `AnyRangeExpression<Bound>` could be a trivial
> closure-based type erasing wrapper. Would it be worth including?

Probably not.  Is there a use-case for this?  Type-erasure is not
usually a win for simple abstractions behind which there is very little
complexity.

dabrahams commented:
on Thu Jul 21 2016, Brent Royal-Gordon <notifications-AT-github.com> wrote:

> Is this a bug? Any idea how I can work around it?

Maybe.  If you provide instructions for reproducing it (or even tell us
what line the error appears on), we might be able to diagnose it.

It's almost certainly a bug that you're seeing any message with a “τ” in
it, FWIW.

-- 
Dave

brentdax commented:
Reproduction is as indicated: copy those two declarations and paste them into a REPL or playground. (Or put them in the standard library, of course, but it reproduces without that.)

dabrahams commented:
Definitely a bug.  I’m looking for workarounds, but not optimistic.  @rudkx @DougGregor is this something that can be easily fixed?

dabrahams commented:
As a workaround, I would try dropping the `relative(to:)` interface and put the interface for completing ranges on `Collection`.

rudkx commented:
I we might be getting confused over the generic parameter ``Bound`` from CountableRange vs. the ``associatedtype Bound`` from RangeExpression.

dabrahams commented:
@rudkx Changing `Bound` to `Bound_` in the snippet doesn’t appear to change anything about the result.

brentdax commented:
Ugly workaround that preserves the general structure of `RangeExpressible`:

```swift
public protocol RangeExpression {
    associatedtype Bound : Comparable
    func relativeTo<BoundDistance: SignedInteger>(start startIndex: Bound, end endIndex: Bound, offset: (Bound, BoundDistance) -> Bound) -> Range<Bound>
}

extension CountableRange: RangeExpression {
    public func relativeTo<BoundDistance: SignedInteger>(start startIndex: Bound, end endIndex: Bound, offset: (Bound, BoundDistance) -> Bound) -> Range<Bound> {
        return Range(self)
    }
}
```

dabrahams commented:
Hey, whatever it takes, man!  Congratulations on finding it!

brentdax commented:
Okay, I'm making forward progress again.

There's a comment in [some relevant code](https://github.com/apple/swift/blob/master/stdlib/public/core/CollectionAlgorithms.swift.gyb#L573) that references a radar I can't read:

```swift
% for Self in 'Indexable', 'MutableIndexable':
[snip]
// WORKAROUND rdar://25214066 - should be on Collection
extension ${Self} {
```

Can someone take a look at that and distill it for me? Currently I'm putting the equivalents of these subscripts on Collection, and I'm wondering if that's going to be a problem.

dabrahams commented:
on Fri Jul 22 2016, Brent Royal-Gordon <notifications-AT-github.com> wrote:

> There's a comment in [some relevant
> code](https://github.com/apple/swift/blob/master/stdlib/public/core/CollectionAlgorithms.swift.gyb#L573)
> that references a radar I can't read:
>
> ```swift
> % for Self in 'Indexable', 'MutableIndexable':
> [snip]
> // WORKAROUND rdar://25214066 - should be on Collection
> extension ${Self} {
> ```
>
> Can someone take a look at that and distill it for me? 

It means I ran into a compiler bug and this was the workaround.  I have
no idea whether the bug is still alive.  I suggest that if you run into
trouble you follow the pattern already in the standard library.

-- 
Dave

rudkx commented:
If I'm not mistaken, the same-type constraint error in your original snippet is due to CountableRange's 'Bound' parameter not currently being required to conform to Comparable.

Either way, the diagnostic here is not useful/actionable, so a new bug at bugs.swift.org would be appreciated.

rudkx commented:
Scratch that, I now see a 'where' clause that requires Bound to be Comparable:
  where
  // WORKAROUND rdar://25214598 - should be just Bound : Strideable
  Bound : _Strideable & Comparable,
  Bound.Stride : SignedInteger {

brentdax commented:
The prototype with `RangeExpression` is now operational (and seems to pass all relevant tests), and I've updated the proposal to match. I'd like opinions on a couple points:

* I've added @dabrahams as a coauthor since `RangeExpression` is his design. I'll take you back off if you don't want to be there (or just take yourself off).

* With further work, the `RangeExpression` workaround method actually ended up with a fairly elegant design:
  ```swift
  range.relative(to: startIndex ..< endIndex, offsettingBy: index(_:offsetBy:))
  func relative<BoundDistance: SignedInteger>(to bounds: Range<Bound>, offsettingBy offset: (Bound, BoundDistance) -> Bound) -> Range<Bound>
  ```
  The proposal calls this temporary and suggests the documentation call it out as likely to change, but should we instead go with this as the intended design?

* Now that we're touching `removeSubrange(i ... j)` and `replaceSubrange(i ... j, with: [x, y])`, I'm wondering if they should be renamed as well, perhaps to one of these:

  * `remove(i ... j)` and `replace(i ... j, with: [x, y])`
  * `remove(at: i ... j)` and `replace(at: i ... j, with: [x, y])`
  * `remove(in: i ... j)` and `replace(in: i ... j, with: [x, y])`

Other than these points, I think this is ready to go.

dabrahams commented:
on Sun Jul 24 2016, Brent Royal-Gordon <notifications-AT-github.com> wrote:

> The prototype with `RangeExpression` is now operational (and seems to
> pass all relevant tests), 

Nice!  IMO you should submit a PR (with a do not merge until proposal is
accepted note) and reference that from the proposal.

> and I've updated the proposal to match. I'd like opinions on a couple
> points:
>
> * I've added @dabrahams as a coauthor since `RangeExpression` is his
> design. I'll take you back off if you don't want to be there (or just
> take yourself off).

Thanks, that's fine.

> * With further work, the `RangeExpression` workaround method actually ended up with a fairly elegant design:
>   ```swift
>   range.relative(to: startIndex ..< endIndex, offsettingBy: index(_:offsetBy:))
>   func relative<BoundDistance: SignedInteger>(to bounds: Range<Bound>,
> offsettingBy offset: (Bound, BoundDistance) -> Bound) -> Range<Bound>
>   ```
>   The proposal calls this temporary and suggests the documentation
> call it out as likely to change, but should we instead go with this as
> the intended design?

I'm really glad you found a workaround, but I don't think it's where we
should end up.  Do you disagree?

> * Now that we're touching `removeSubrange(i ... j)` and
> `replaceSubrange(i ... j, with: [x, y])`, I'm wondering if they should
> be renamed as well, perhaps to one of these:
>
>   * `remove(i ... j)` and `replace(i ... j, with: [x, y])`
>   * `remove(at: i ... j)` and `replace(at: i ... j, with: [x, y])`
>   * `remove(in: i ... j)` and `replace(in: i ... j, with: [x, y])`
>
> Other than these points, I think this is ready to go.

Those names were very carefully chosen, and I guess I'd rather not
consider changing them as part of this proposal, which is mostly
orthogonal.

rudkx commented:
FYI I created a test case that doesn't rely on the stdlib as it exists today, and opened a JIRA bug for it: https://bugs.swift.org/browse/SR-2158

brentdax commented:
> I'm really glad you found a workaround, but I don't think it's where we should end up.  Do you disagree?

I think you said you preferred the name `RangeExpression` over something involving "subrange" because you wanted it to be more broadly applicable than that. On the other hand, the mere fact that this API is more broadly applicable than one that uses `Collection` doesn't necessarily mean this is the more-broadly-applicable API we want, so it's probably best to keep our options open for now.

With this last batch of minor tweaks, I think this is ready to go. @dabrahams, @lattner, it's your show.

@rudkx Thank you! I meant to fully reduce it once we got this thing merged, but you've saved me the effort. I've updated the comments in the prototype to point to that bug.

jtbandes commented:
Is there any concern that `prefix operator ...` might conflict with the syntax for some future parameter-pack-ish feature?

brentdax commented:
@jtbandes That's an interesting point, and one I hadn't considered.

I've chosen to propose the full set of operators that would make sense for this feature and allow the reviewers and core team to whittle them down if they want something narrower. I won't be upset if some of them are removed.

(For what it's worth, postfix `...` is the most expendable operator—it's included for completeness, but it's not actually useful for subscripting. The infix forms provide access to all of this proposal's functionality, at the cost of having to write—and more importantly, to read and interpret—`array[i..<nil]` instead of `array[i..<]`.)
